File: August 2022/August_19.cpp
class Solution {
public:
    bool isPossible(vector<int>& nums) {
        priority_queue<int, vector<int>, greater<int> > pq;
        int i = 0, n = nums.size();
        while(i < n && nums[i] == nums[0]){
            pq.push(1);
            i++;
        }
        while(i < n){
            int j = i;
            vector<int> vec;
            while(j < n && nums[j] == nums[i - 1] + 1){
                if(!pq.empty()){
                    int rem = pq.top();
                    pq.pop();
                    vec.push_back(rem + 1);
                } else
                    vec.push_back(1);
                j++;
            }
            if(i == j)
                vec.push_back(1);
            i = max(j, i + 1);
            if(!pq.empty() && pq.top() < 3)
                return false;
            while(!pq.empty())
                pq.pop();
            for(int &v : vec)
                pq.push(v);
        }
        if(!pq.empty() && pq.top() < 3)
            return false;
        return true;
    }
};

File: August 2022/August_19.java
class Solution {
    public boolean isPossible(int[] nums) {
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        pq.add(1);
        int i = 1, n = nums.length;
        while(i < n && nums[i] == nums[0]){
            pq.add(1);
            i++;
        }
        while(i < n){
            int j = i;
            List<Integer> list = new ArrayList<>();
            while(j < n && nums[j] == nums[i - 1] + 1){
                if(!pq.isEmpty())
                    list.add(pq.remove() + 1);
                else
                    list.add(1);
                j++;
            }
            if(j == i)
                list.add(1);
            i = Math.max(j, i + 1);
            if(!pq.isEmpty() && pq.peek() < 3)
                return false;
            pq.clear();

            for(int v : list)
                pq.add(v);
        }
        if(!pq.isEmpty() && pq.peek() < 3)
            return false;
        return true;
    }
}
